packages/core/content-manager/server/src/history/services/utils.ts

```diff
@@ -1,42 +1,234 @@
+          previousAttributesObject[diffKey] = source[diffKey];
+
+          return previousAttributesObject;
+        },
+        {}
+      );
+    };
+
+    const versionSchemaKeys = Object.keys(versionSchemaAttributes);
+    const contentTypeSchemaAttributesKeys = Object.keys(sanitizedContentTypeSchemaAttributes);
+    // The attribute is new if it's on the content type schema but not on the version schema
+    const uniqueToContentType = difference(contentTypeSchemaAttributesKeys, versionSchemaKeys);
+    const added = reduceDifferenceToAttributesObject(
+      uniqueToContentType,
+      sanitizedContentTypeSchemaAttributes
+    );
+    // The attribute was removed or renamed if it's on the version schema but not on the content type schema
+    const uniqueToVersion = difference(versionSchemaKeys, contentTypeSchemaAttributesKeys);
+    const removed = reduceDifferenceToAttributesObject(uniqueToVersion, versionSchemaAttributes);
+
+    return { added, removed };
+  };
+
+  /**
+   * @description
+   * Gets the value to set for a relation when restoring a document
+   */
+  const getRelationRestoreValue = async (
+    versionRelationData: Data.Entity,
+    attribute: Schema.Attribute.RelationWithTarget
+  ) => {
+    if (Array.isArray(versionRelationData)) {
+      if (versionRelationData.length === 0) return versionRelationData;
+
+      const existingAndMissingRelations = await Promise.all(
+        versionRelationData.map((relation) => {
+          return strapi.documents(attribute.target).findOne({
+            documentId: relation.documentId,
+            locale: relation.locale || undefined,
+          });
+        })
+      );
+      const existingRelations = existingAndMissingRelations.filter(
+        (relation) => relation !== null
+      ) as Modules.Documents.AnyDocument[];
+
+      return existingRelations;
+    }
+    const existingRelationOrNull = await strapi.documents(attribute.target).findOne({
+      documentId: versionRelationData.documentId,
+      locale: versionRelationData.locale || undefined,
+    });
+
+    return existingRelationOrNull;
+  };
+
+  /**
+   * @description
+   * Gets the value to set for a media asset when restoring a document
+   */
+  const getMediaRestoreValue = async (
+    versionRelationData: Data.Entity,
+    attribute: Schema.Attribute.Media<any, boolean>
+  ) => {
+    if (attribute.multiple) {
+      const existingAndMissingMedias = await Promise.all(
+        // @ts-expect-error Fix the type definitions so this isn't any
+        versionRelationData.map((media) => {
+          return strapi.db.query('plugin::upload.file').findOne({ where: { id: media.id } });
+        })
+      );
+
+      const existingMedias = existingAndMissingMedias.filter((media) => media != null);
+      return existingMedias;
+    }
+
+    const existingMediaOrNull = await strapi.db
+      .query('plugin::upload.file')
+      .findOne({ where: { id: versionRelationData.id } });
+
+    return existingMediaOrNull;
+  };
```

Comment:
ℹ️  The diff doesn't make it easy to see that these are the two functions created to handle restoring relations and media respectively